### ISSUE TITLE
Use of deprecate PyUnicode_FromUnicode.

### ISSUE THREAD

**Issue by methane** [cython/cython#3678]
I am searching use of deprecated API and I found PyUnicode_FromUnicode is used in spaCy.
I confirmed Cython 0.29.20 produce same code.

Cython source:
https://github.com/explosion/spaCy/blob/bab30e4ad2ad35d7133b7f8027a3558a02e018e4/spacy/pipeline/pipes.pyx#L1441-L1451

Generated code:
```py
  /* "spacy/pipeline/pipes.pyx":1446
 *             '', '', '', '', '', '', '', '', '', '', '', '', '',
 *             '', '', '', '', '', '', '', '', '', '', '', '', '',
 *             '', '', '', '', '', '', '', '', '', '', '', '',             # <<<<<<<<<<<<<<
 *             '', '', '', '', '', '', '', '', '', '', '', '', '',
 *             '', '', '', '', '', '', '', '', '', '', '', '', '',
 */
  __pyx_ustring__374 = PyUnicode_FromUnicode(__pyx_k__375, (sizeof(__pyx_k__375) / sizeof(Py_UNICODE))-1); if (unlikely(!__pyx_ustring__374)) __PYX_ERR(0, 144
  if (__Pyx_PyUnicode_READY(__pyx_ustring__374) < 0) __PYX_ERR(0, 1446, __pyx_L1_error)
  __pyx_ustring__376 = PyUnicode_FromUnicode(__pyx_k__377, (sizeof(__pyx_k__377) / sizeof(Py_UNICODE))-1); if (unlikely(!__pyx_ustring__376)) __PYX_ERR(0, 144
  if (__Pyx_PyUnicode_READY(__pyx_ustring__376) < 0) __PYX_ERR(0, 1446, __pyx_L1_error)
  __pyx_ustring__378 = PyUnicode_FromUnicode(__pyx_k__379, (sizeof(__pyx_k__379) / sizeof(Py_UNICODE))-1); if (unlikely(!__pyx_ustring__378)) __PYX_ERR(0, 144
  if (__Pyx_PyUnicode_READY(__pyx_ustring__378) < 0) __PYX_ERR(0, 1446, __pyx_L1_error)
  __pyx_ustring__380 = PyUnicode_FromUnicode(__pyx_k__381, (sizeof(__pyx_k__381) / sizeof(Py_UNICODE))-1); if (unlikely(!__pyx_ustring__380)) __PYX_ERR(0, 144
  if (__Pyx_PyUnicode_READY(__pyx_ustring__380) < 0) __PYX_ERR(0, 1446, __pyx_L1_error)
  __pyx_ustring__382 = PyUnicode_FromUnicode(__pyx_k__383, (sizeof(__pyx_k__383) / sizeof(Py_UNICODE))-1); if (unlikely(!__pyx_ustring__382)) __PYX_ERR(0, 144
  if (__Pyx_PyUnicode_READY(__pyx_ustring__382) < 0) __PYX_ERR(0, 1446, __pyx_L1_error)
  __pyx_ustring__384 = PyUnicode_FromUnicode(__pyx_k__385, (sizeof(__pyx_k__385) / sizeof(Py_UNICODE))-1); if (unlikely(!__pyx_ustring__384)) __PYX_ERR(0, 144
  if (__Pyx_PyUnicode_READY(__pyx_ustring__384) < 0) __PYX_ERR(0, 1446, __pyx_L1_error)
  __pyx_ustring__386 = PyUnicode_FromUnicode(__pyx_k__387, (sizeof(__pyx_k__387) / sizeof(Py_UNICODE))-1); if (unlikely(!__pyx_ustring__386)) __PYX_ERR(0, 144
  if (__Pyx_PyUnicode_READY(__pyx_ustring__386) < 0) __PYX_ERR(0, 1446, __pyx_L1_error)
```

But I can not find how Cython produce this code block.
I'm sorry if this code block is not generated by Cython.

**Comment by scoder** [comment]
Probably here:
https://github.com/cython/cython/blob/7f26ba41ced0a95e2aef6ed608302fe743d836dd/Cython/Compiler/ExprNodes.py#L1693

Not sure if the codec issue (from Py3.3) still persists. Probably not. Worth revisiting.

**Comment by scoder** [comment]
I think the condition is wrong, though. This is only needed for _lone_ surrogates, not for normal non-BMP characters (which result in surrogate pairs).

**Comment by scoder** [comment]
Ok, the condition wasn't wrong. On a Py2 16bit Unicode platform, it is impossible to distinguish a 1-character Unicode string with a non-BMP character in it from a 2-character Unicode string with the two corresponding surrogate codes in it. Translating such source code in a 16bit Unicode runtime would give a different result than on a 32bit runtime, because we are currently using regular Python Unicode strings for the internal string values.

Two solutions come to mind:
- replace the internal Unicode string representation to allow it to safely represent this case
- wait with the fix until Cython 3.1 which drops Py2 support and can thus expect the full Unicode range to be available.

**Comment by scoder** [comment]
#3697 avoids the use of `PyUnicode_FromUnicode()` unless actual (lone) surrogates are used. We still have to find a way to make that work, but it's a much rarer use case than non-BMP characters in strings.

While implementing tests, I stumbled over an interesting Py2.7 bug where `u"\uD800\uDFFF"` was treated differently the first time the Python module is loaded than in subsequent imports. I guess it's an issue with the way Unicode string literals are serialised in `.pyc` files.

I'll not file a bug report for that now. ;-) It also doesn't affect the processing in Cython since it only applies to literals, not generated strings.

**Comment by scoder** [comment]
Fixed as good as it gets in #3697. We still use the old functions in the code, but only a) if we find lone surrogates in a Unicode string or b) when we generate code for non-BMP unicode strings on a 16-bit Python 2.x deployment (including Windows).

This is the first time we do this and the previously generated code was intentionally left like that in order to avoid this difference. Let's hope that most users generate their code in Py3 by now.
